### ISSUE TITLE
Freqtrade causing my VPS to crash

### ISSUE THREAD

**Issue by King-Techno** [freqtrade/freqtrade#3016]
Everytime I run freqtrade on my VPS (EC2 server) it works for an hour or two. Then it causes my VPS to crash and stop working. I honestly do not know what to do.

I am using Ubuntu 18.04 and my VPS has 1 CPU.

**Comment by xmatthias** [comment]
well what do the logs show?
how much ram have you got?

**Comment by King-Techno** [comment]
I dont know how to check the logs?
I installed freqtrade using docker.

VPS Data:
RAM: 1.0 GB, CPU: 1.

The issue started after I changed my strategy. I suspect that my CPU is too low for my current strategy and I need to upgrade my VPS.

**Comment by King-Techno** [comment]
The issue occurred after I added this to my strategy `dataframe['high'].rolling(period).max().shift()`

**Comment by xmatthias** [comment]
You should be able to check the logs by using journalctl (you'll need to investigate parameters to get messages from the previous boot - probably it's `journalctl -b -1` but i'm not certain at the moment.

Honestly, this is a no-brainer for me.

Minimal requirements are stated as 2GB Ram, 2vCPU - you're not filling either of these.

Also, requirements depend on the strategy - so if you're using a ton of indicators, 2GB may not be enough. If it's compute-intensive indicators, 2vCPU may not be enough....

**Comment by King-Techno** [comment]
I agree as well that I will need to upgrade my server. I give that a shot and hopefully, it should work.

**Comment by King-Techno** [comment]
I think the issue may be with freqtrade. I am having sudden spikes in my CPU usage causing me to use up all my credits. 

![image](https://user-images.githubusercontent.com/53098717/75784369-05903c80-5d5a-11ea-9f99-10bce707d909.png)

**Comment by xmatthias** [comment]
From the timing perspective, i would assume you're using VolumePairlist - which is usually running every 6 hours.
However - a graph without specification as to what this is (memory / cpu / network) does not say anything.

**Comment by King-Techno** [comment]
Sorry about that @xmatthias . I have made alterations to my server, so I will wait a few hours to see if there are any improvements. In the meantime I will close the case and reopen if issues persist. Thank you for your time though.

**Comment by King-Techno** [comment]
## Step 1: Have you search for this issue before posting it?

Yes

If you have discovered a bug in the bot, please [search our issue tracker](https://github.com/freqtrade/freqtrade/issues?q=is%3Aissue). 
If it hasn't been reported, please create a new issue.

## Step 2: Describe your environment

  * Operating system: ubuntu
  * Python Version: 3.6.9 (`python -V`)
  * CCXT version: _____ (`pip freeze | grep ccxt`)
  * Branch: Develop
  * Last Commit ID: no git repository (`git log --format="%H" -n 1`)
  
## Step 3: Describe the problem:

I am my CPU for my virtual machine is spiking to 100%. This is casuing my virtual machine is break and stop working or slow down a lot. 

### Steps to reproduce:

  1. Run freqtrade on my virtual machine
  2. Wait a few hours and the issue starts.
 
Please note this issue only started when I altered my strategy. 
  
### Observed Results:

I know what is causing my CPU to reach 100%:
-My free memory is rapidly declining. This means that something I am running on reqtrade is draining my VPS memory. Maybe my database is corrupt?


### Relevant code exceptions or logs:

  ```
  // paste your log here
  ```

I cant access my logs. I have tried using `journalctl -b -1`

**Comment by King-Techno** [comment]
So after some research, I have found the issue! There is a memory leak when .rolling().max() is used. 

Can refer to here: https://github.com/pandas-dev/pandas/issues/25893 

This means I may just have to alter my strategy. Are there any alternatives for this?

**Comment by hroff-1902** [comment]
that bug was fixed in pandas an year ago, as one can find from that issue at the pandas repo...

**Comment by King-Techno** [comment]
hmmm then I am not sure what is causing the issue. I must have something to do with `.rolling().max()`  as it only started when I added that to my strategy. Any suggestions on what I can do?

**Comment by hroff-1902** [comment]
which version of pandas is installed on the machine?

but I agree with @xmatthias , due to timings (every 6 hrs) this can be related to VolumePairList used... further investigations may be needed in this case...

as a workaround, you may try to limit your pairlist -- define some shorter list of pairs in the whitelist setting in the config and add StaticPairList before VolumePairList (or StaticPairList instead of VolumePairList)

**Comment by King-Techno** [comment]
How do i check the version of pandas installed on the machine?

Also in regards to the VolumePairList, I am actually only using StaticPairLists alongside the pairs in whitelist.

**Comment by hroff-1902** [comment]
`pip3 list|grep pandas`
this should be run inside the docker, if freqtrade runs in the docker...

**Comment by hroff-1902** [comment]
do you have these spikes of cpu usage always every 6 hrs as on the picture above or they are random and 6 hrs on the picture is only one example with 6 hrs interval randomly chosen?

**Comment by King-Techno** [comment]
I logged into docker on my terminal then run the code. I don't actually have pandas installed on my machine. 

Also it happens 1-2 hours after I start my machine. It doesn't occur every 6 hours.

**Comment by hroff-1902** [comment]
ah, there can be no pip inside the docker installed... let's wait for @xmatthias, I'm not very familiar with docker...

**Comment by xmatthias** [comment]
docker is just a container around a regular linux (debian) image ... of course there is pip installed.

Are you using the official image? as then i'm certain there is pandas 1.0.x installed - at least if you're using a recent image.


To debug this issue - best install htop (it's a "better" version of top) ... then filter on memory - and observe (hopefully you're looking while memory spikes ...).
Maybe it's something else entirely?

If it happesn after 1-2 hours of running freqtrade, it's not the Dynamic pairlist - that runs every 6 hours (and would run right at startup of the bot).

---
Just as an example:
I've used oracle cloud (for a short time since it was/is free) - and they have a daemon to get the metrics for the UI. it did consume >50% of all available memory on that machine ... after switching it off i'm able to run ~3 bots in parallel on their smallest available machine (which i think has 500MB...).

Now i'm not saying that's reccomended, nor would i do that for a bot running live, but for dry-runs it's ok - and buys/sells between that dry-bot, and a live bot running the same strategy on a more powerfull machine were aligned (that bot runs for weeks without any issues / memory spikes).

If you discover that the memory issue is freqtrade, then best very carefully look at your strategy - are you assigning things to `self.xxx` to keep track of? 
That's one example - but there's others where a strategy can easily cause a memory leak.

**Comment by King-Techno** [comment]
Thank you @xmatthias

Yes I am using the image provided in the docker installation guide provided by freqtrade. 

I am going to try that. I have removed .`rolling().max().shift `from my strategy and I am going to use the method you suggested to track freqtrade as it runs. If my machine doesn't fail I will know that the addition of `.rolling().max().shift ` is the issue. so far things seem stable but i will give it a couple hours

**Comment by King-Techno** [comment]
@xmatthias I can confirm that the issue is when `.rolling().max().shift` is added to me strategy. My machine has been running perfectly since I changed my strategy. Do you have any suggestions as to why this may occur? I think it is a bug.

**Comment by King-Techno** [comment]
I have a feeling the issue may be because I never specified my `.shift()` with a number e.g.`1`. I think doing this may cause some kind of reoccurring loop which is draining the memory. What do you think?

**Comment by xmatthias** [comment]
`.shift()` defaults to 1 if no parameter is specified. Feel free to test it with adding 1, but i don't think that will make a difference.

Either way, this confirms my suspicion, it's related to your strategy, not direcly to freqtrade.

**Comment by hroff-1902** [comment]
`.shift()` is equal to `.shift(periods=1)` or `.shift(1)`

**Comment by xmatthias** [comment]
may i assume that you've found a solution to this?

**Comment by King-Techno** [comment]
I still haven’t found the solution unfortunately

**Comment by xmatthias** [comment]
Honestly, i struggle to see how we can help you - it's a problem within strategy code, and a problem in pandas `.shift()` method is highly unlikely.

I suspect you're doing something else - My suspicion is that you're storing something in the strategy class  - that's the most obvious problem i can think of.

**Comment by King-Techno** [comment]
I’ll give it another review. For now I’ll close this issue. Thanks for the help though.

**Comment by fgiroult321** [comment]
Not sure i can comment as the issue is closed, but i am facing the exact same issue @King-Techno described.
I started using a strategy with a rolling min and i observe the free memory on my server decreasing. After a couple of hours it used all available memory and i can't connect to the machine anymore.
To make sure i upgraded my EC2 instance to a t3.large with 8GB of RAM - freqtrades eats all the available memory after 12 hours. 
I also tested using docker on an AWS AMI, and using freqtrade script installation on an Ubuntu 18.04 - it behaves the same.
I will be happy to provide specific log if needed.

**Comment by xmatthias** [comment]
@fgiroult321 the error you're having is similar to #2979 - which is caused by https://github.com/pandas-dev/pandas/issues/32266 - a memory-error in pandas when using `.rolling().max()` (or .min()`.

**Comment by fgiroult321** [comment]
Thanks @xmatthias for the links - so i guess downgrading to pandas 0.25.x or using the workaround should do the trick

**Comment by xmatthias** [comment]
well yes - i did downgrade pandas to 0.25 to verify, and it seems to work just fine - and we're not using any of the new features, so i think that'll be the best / easiest / most performant method.

**Comment by fgiroult321** [comment]
i can confirm that, after few minutes running my strategy with pandas 0.25.3, memory usage is steady and reasonably low. Thanks again for your support!

**Comment by King-Techno** [comment]
As suspected... the issue wasn't with my strategy. Thanks for rasing this up again @fgiroult321

**Comment by fgiroult321** [comment]
@King-Techno : i was desperate not being able to use my strategy that works quite well ;)

**Comment by fgiroult321** [comment]
Quick update : panda 1.0.4 fixes the issue - no more memory leaks using .rolling().max()
Thanks

**Comment by Chyang777** [comment]
hello @fgiroult321 , how can I downgrade the pandas, could I know how to do that, thank you. I'm using docker.

**Comment by fgiroult321** [comment]
Hi, you need to specify the panda version you want in the requirements.txt file and rebuild the docker image
